File: Cpp/K08_2sub.cpp
#include <iostream>
#include <stdexcept>
#include "Golem.h"
#include "K08_2sub.h"

K08_2sub::K08_2sub()
{
}

K08_2sub::~K08_2sub()
{
}

enum AttackMethod {
	Attack = 65,
	Skill = 99,
	Magic = 133
};

void K08_2sub::AttackGolem(int lv, int& playerHp)
{
	try {
		Golem golem = Golem(lv * 50 + 100, lv * 10 + 40, lv * 10 + 30);

		std::cout << "ゴーレムLv." << lv + 1 << "が現れた！\n";
		while (golem.hp > 0)
		{
			std::cout << "\n残りHP：" << golem.hp << "\n";
			std::cout << "武器を選択してください（１．攻撃　２．特技　３．魔法）＞";
			int choice = 0;
			std::cin >> choice;
			int damage = 0;

			switch (choice)
			{
			case 1:
				damage = static_cast<int>(Attack);
				break;
			case 2:
				damage = static_cast<int>(Skill);
				break;
			case 3:
				damage = static_cast<int>(Magic);
				break;
			default:
				break;
			}
			damage -= golem.defense;

			if (damage <= 0) {
				damage = 0;
				std::cout << "ゴーレム：「ハハハハハ、情けないな！貴様は弱すぎる！」\n";
				std::cout << "ゴーレムがあなたを攻撃しました！攻撃値：" << golem.attack << "\n";
				playerHp -= golem.attack;
				if (playerHp <= 0)
				{
					playerHp = 0;
					std::cout << "あなたはゴーレムに負けました！ゲームオーバー！\n";
					return;
				}
				std::cout << "あなたの残りHPは：" << playerHp << "\n";
			}
			std::cout << "ダメージは" << damage << "です。\n";
			golem.hp -= damage;

			if (golem.hp <= 0)
				golem.hp = 0;
			if (golem.hp == 0)
				std::cout << "ゴーレムLv." << lv + 1 << "を倒した！\n";
		}
	}
	catch (const std::exception& ex) {
		std::cerr << ex.what();
	}
}

File: CSharp/k08_2sub.cs
using System;

namespace CSharp
{
    static partial class Utilities
    {
        enum AttackMethod
        {
            Attack = 65,
            Skill = 99,
            Magic = 133
        }

        public static void AttackGolem(int lv, ref int playerHp)
        {
            try
            {
                Golem golem = new Golem(
                    hp: lv * 50 + 100,
                    defense: lv * 10 + 40,
                    attack: lv * 10 + 30
                    );

                Console.WriteLine(string.Format("ゴーレムLv.{0}が現れた！", lv + 1));
                while (golem.Hp > 0)
                {
                    Console.WriteLine(string.Format("\n残りHP：{0}", golem.Hp));
                    Console.Write("武器を選択してください（１．攻撃　２．特技　３．魔法）＞");
                    int choice = int.Parse(Console.ReadLine());
                    int damage;
                    switch (choice)
                    {
                        case 1:
                            damage = (int)AttackMethod.Attack;
                            break;
                        case 2:
                            damage = (int)AttackMethod.Skill;
                            break;
                        case 3:
                            damage = (int)AttackMethod.Magic;
                            break;
                        default:
                            damage = 0;
                            break;
                    }

                    damage -= golem.Defense;
                    if (damage <= 0)
                    {
                        damage = 0;
                        Console.WriteLine("ゴーレム：「ハハハハハ、情けないな！貴様は弱すぎる！」");
                        Console.WriteLine(string.Format("ゴーレムがあなたを攻撃しました！攻撃値：{0}", golem.Attack));
                        playerHp -= golem.Attack;
                        if (playerHp <= 0)
                        {
                            playerHp = 0;
                            Console.WriteLine("あなたはゴーレムに負けました！ゲームオーバー！");
                            return;
                        }
                        Console.WriteLine(string.Format("あなたの残りHPは：{0}", playerHp));
                    }
                    Console.WriteLine(string.Format("ダメージは{0}です。", damage));
                    golem.Hp -= damage;

                    if (golem.Hp <= 0)
                        golem.Hp = 0;

                    if (golem.Hp == 0)
                        Console.WriteLine(string.Format("ゴーレムLv.{0}を倒した！", lv + 1));
                }
            }
            catch (Exception ex)
            {
                Console.Error.WriteLine(ex.Message);
            }
        }
    }
}
